devito/visitors.py

```diff
@@ -363,9 +414,55 @@ def visit_Iteration(self, o, subs={}, offsets={}):
             subs[o.dim] = Symbol(vname)
             return o._rebuild(*nodes, index=vname)
 
-    def visit_Expression(self, o, subs={}, offsets={}):
-        # Collect all offsets used with a dimension
-        # TODO: Straight update is not really safe,
-        # since we're mapping into a list...
-        offsets.update(o.index_offsets)
+    def visit_Expression(self, o, subs={}, offsets=defaultdict(set)):
+        """Collect all offsets used with a dimension"""
+        for dim, offs in o.index_offsets.items():
+            offsets[dim].update(offs)
         return o
+
+
+class MergeOuterIterations(Transformer):
+    """
+    :class:`Transformer` that merges subsequent :class:`Iteration`
+    objects iff their dimenions agree.
+    """
+
+    def is_mergable(self, iter1, iter2):
+        return (iter1.dim == iter2.dim and
+                iter1.index == iter2.index and
+                iter1.limits == iter2.limits)
```

Comment:
`offsets` are used to construct `limits` so they are implied - but yes, I should leave a comment to explain that.